modules/dnn/src/layers/slice_layer.cpp

```diff
@@ -258,6 +297,28 @@ class SliceLayerImpl : public SliceLayer
         CV_Assert(inputs.size() == 1);
         const MatSize& inpShape = inputs[0].size;
 
+        if (hasSteps && hasNegSteps)
+        {
+            int start, end, new_start, new_end, step;
+            for (int i = 0; i < sliceSteps[0].size(); ++i)
+            {
+                step = sliceSteps[0][i];
+                if (step < 0)
+                {
+                    step = -step;
+                    start = sliceRanges[0][i].start;
+                    end = sliceRanges[0][i].end;
+                    new_end = start >= inpShape[i] ? inpShape[i] : start + 1;
+                    new_start = end + 1;
+                    new_start = new_end - 1 - ((new_end - 1 - new_start) / step) * step;
+
+                    sliceSteps[0][i] = step;
+                    sliceRanges[0][i].start = new_start;
+                    sliceRanges[0][i].end = new_end;
+                }
+            }
+        }
```

Comment:
Copypaste?

Reply:
Yes, it is needed since we cannot change values in `getMemoryShapes`. The reason why I want to change the values after the constructor is I want to preserve the original starts, ends and steps, so that we might be able to support other backends if they support negative steps.

We can avoid this if `cv::Range` can support steps and negative steps, but that should be another pull request.

Reply:
Done with putting those in a function.

> The code is diverging already: here you didn't negate the condition.

I don't quite understand here. As `i` keeps incrementing itself, why the code can go diverging?

Reply:
Copypasted code was not in sync already.